### ISSUE TITLE
org.eclipse.equinox.p2.discovery 1.2.0.20210315-2042 considered unsigned (has jarsigner)

### ISSUE THREAD

**Issue by mickaelistria** [eclipse-equinox/p2#74]
Trying to install Eclipse MPC from https://download.eclipse.org/mpc/nightly/ in a fresh Eclipse SDK I-Builds shows me a popup about org.eclipse.equinox.p2.discovery 1.2.0.20210315-2042 being unsigned. However, the jar has jarsigner files in it.

**Comment by mickaelistria** [comment]
I also see this issue when installing LSP4E; complaining about LSP4J and Guava bundles being unsigned, although they seem to have jarsigner info available.

**Comment by merks** [comment]
Perhaps this is related to all the changes to use different OSGi libraries given that this issue appears to be new...

**Comment by mickaelistria** [comment]
I also just got it for snakeyaml (from Orbit), while installing TM4E

> Perhaps this is related to all the changes to use different OSGi libraries given that this issue appears to be new...

Maybe. I believe that some jarsigner info are not recognized or are ignored. Could it be some expired signer certificate or unsupported algorithm making jarsigner assume it's not relevant here?

**Comment by merks** [comment]
Well, I've been investigating the fact that nothing actually validates that the signatures are within the validity range of the certificate, but I've not seen this behavior.  Given me a few minutes to update my TP of the Oomph IDE to the latest and try to do a fresh install into an empty user environment.  That should reproduce such a problem...

**Comment by laeubi** [comment]
> 1.2.0.20210315-2042

From the timestamp this seems to be > 1 year old artifact, maybe the signature is no longer valid eg. expired?

**Comment by merks** [comment]
Oh, and given that mostly this jar signing validation is now delegated to the JRE implementation, it's probably highly relevant which JRE/JDK you are using.

@laeubi 

Much of the content is signed with expired certificates.

https://download.eclipse.org/oomph/archive/reports/download.eclipse.org/eclipse/updates/4.24-I-builds/index.html

But only one is actually invalid where the signature was produced after the certificate expired (and as I suggested, nothing in Equinox or p2 validates that).

**Comment by laeubi** [comment]
> Oh, and given that mostly this jar signing validation is now delegated to the JRE implementation

But the JRE might validate this already, e.g. an expired certificate is not bad as long as it was time-stamped, but if not the JRE might still complain?

**Comment by laeubi** [comment]
@mickaelistria can you try manually verify the jar in question?

https://docs.oracle.com/javase/tutorial/deployment/jar/verify.html

**Comment by laeubi** [comment]
> it's probably highly relevant which JRE/JDK you are using.

Even to the settings:

> The JDK treats a signed JAR file as unsigned if the signed JAR file uses any algorithm that's specified in the jdk.jar.disabledAlgorithms Security Property in the java.home/lib/security/java.security file (where java.home is the directory where you installed your JRE).

**Comment by mickaelistria** [comment]
> Oh, and given that mostly this jar signing validation is now delegated to the JRE implementation, it's probably highly relevant which JRE/JDK you are using.

OK, it's a good clue. Maybe the newer/latest JRE I used has removed support for some unsafe algorithms that were used for those bundles.
FWIW, here is what I'm currently using
```
$ java -version
openjdk version "18" 2022-03-22
OpenJDK Runtime Environment 22.3 (build 18+37)
OpenJDK 64-Bit Server VM 22.3 (build 18+37, mixed mode, sharing)
```
Listing the bundles in Installation Details say they're not signed.

I usually use JustJ, and want to use it anyway, so let's try installing JustJ (17.0.2 from  https://download.eclipse.org/justj/jres/latest/ ) in this same IDE... and now Installation Details say that the artifacts are signed!

So that confirms a few things:

- p2 seems OK (let's close)
- The issue comes when using Java 18

And brings a new issue: **Some signed bundle will be treated as unsigned with Java 18** I'm not sure what's the best place to report it?

**Comment by merks** [comment]
Note that the current certificate is about to expire:

![image](https://user-images.githubusercontent.com/208716/168265391-136939b3-8f76-48b7-8b5a-3cb5a72f4a49.png)

So let's hope the IT staff get a new one in place before we start producing content signed with an expired certificate!

I'm not sure what we should report for this Java 18 thing?!  What action should be taken?  Force all content to be signed with a newer certificate?

**Comment by mickaelistria** [comment]
> What action should be taken? Force all content to be signed with a newer certificate?

The immediate action would be that infra team sets up signing service with a certificate that is using an up-to-date algorithm that's not deprecated by latest Java version (cc @fgurr @mbarbero ) so the problem doesn't expand.
As for what's already built, I don't think it's a good thing to force a rebuild. We maybe just need to inform everyone so they become aware of the problem and can decide on their end whether to rebuild or not.
Also, if this problem seems critical, we could lock SimRel to use JustJ 17 for the next release at least.

**Comment by merks** [comment]
@jonahgraham FYI  At some point when people install/use Java 18, they'll encounter problems like this with older signed content.

**Comment by jonahgraham** [comment]
We could pgp sign all of simrel?

**Comment by merks** [comment]
This flag suggests one could sign jar signed things:

https://www.eclipse.org/tycho/sitedocs/tycho-gpg-plugin/sign-p2-artifacts-mojo.html#skipIfJarsigned

**Comment by akurtakov** [comment]
> This flag suggests one could sign jar signed things:
> 
> https://www.eclipse.org/tycho/sitedocs/tycho-gpg-plugin/sign-p2-artifacts-mojo.html#skipIfJarsigned

If we do that why jarsign at all?

**Comment by fredg02** [comment]
> So let's hope the IT staff get a new one in place before we start producing content signed with an expired certificate!

@merks please create a HelpDesk issue so we can track this.

**Comment by fredg02** [comment]
> The immediate action would be that infra team sets up signing service with a certificate that is using an up-to-date algorithm that's not deprecated by latest Java version (cc @fredg02 @mbarbero ) so the problem doesn't expand.

I assume that concerns the certificate that is about to expire in a few days anyway, or are other certificates affected as well?

**Comment by merks** [comment]
@fredg02 

I created this.

https://gitlab.eclipse.org/eclipsefdn/helpdesk/-/issues/1292

Note that the current signing certificate is not one that is using an older algorithm that's causing problem.  The one causing the problem described in this issue is this significantly older one:

![image](https://user-images.githubusercontent.com/208716/168288929-f5d3845f-8baf-4b37-aa29-89d77318cb59.png)
